common.go

```diff
@@ -81,35 +84,60 @@ func detachedSignatureInputFromHash(plaintextAndHeaderHash []byte) []byte {
 	return buf.Bytes()
 }
 
-func hmacSHA512256(key []byte, input []byte) []byte {
-	// Equivalent to crypto_auth, but using Go's builtin HMAC. Truncates
-	// SHA512, instead of actually calling SHA512/256.
-	if len(key) != cryptoAuthKeyBytes {
-		panic("Bad crypto_auth key length")
+func copyEqualSize(out, in []byte) {
+	if len(out) != len(in) {
+		panic(fmt.Sprintf("len(out)=%d != len(in)=%d", len(out), len(in)))
 	}
-	authenticatorDigest := hmac.New(sha512.New, key)
-	authenticatorDigest.Write(input)
+	copy(out, in)
+}
+
+func copyEqualSizeStr(out []byte, in string) {
+	if len(out) != len(in) {
+		panic(fmt.Sprintf("len(out)=%d != len(in)=%d", len(out), len(in)))
+	}
+	copy(out, in)
+}
+
+type macKey [cryptoAuthKeyBytes]byte
+
+type payloadHash [sha512.Size]byte
+
+type payloadAuthenticator [cryptoAuthBytes]byte
+
+func (pa payloadAuthenticator) Equal(other payloadAuthenticator) bool {
+	return hmac.Equal(pa[:], other[:])
+}
+
+func computePayloadAuthenticator(macKey macKey, payloadHash payloadHash) payloadAuthenticator {
+	// Equivalent to crypto_auth, but using Go's builtin HMAC. Truncates
+	// SHA512, instead of calling SHA512/256, which has different IVs.
+	authenticatorDigest := hmac.New(sha512.New, macKey[:])
+	authenticatorDigest.Write(payloadHash[:])
 	fullMAC := authenticatorDigest.Sum(nil)
-	return fullMAC[:cryptoAuthBytes]
+	var auth payloadAuthenticator
```

Comment:
I wonder if it would be nice to have some hardcoded `array32FromSlice` functions, which we could call in these situations with a cast, to avoid needing all these temporary buffer variables? I feel like every so often I wish we had those.